hack/generator/pkg/astmodel/conversion_function_builder.go

```diff
@@ -0,0 +1,494 @@
+	return &ConversionFunctionBuilder{
+		IdFactory:             idFactory,
+		CodeGenerationContext: codeGenerationContext,
+		conversions: []ConversionHandler{
+			// Complex wrapper types checked first
+			IdentityConvertComplexOptionalProperty,
+			IdentityConvertComplexArrayProperty,
+			IdentityConvertComplexMapProperty,
+
+			// TODO: a flip function of some kind would be kinda nice (for source vs dest)
+			IdentityAssignValidatedTypeDestination,
+			IdentityAssignValidatedTypeSource,
+			IdentityAssignPrimitiveType,
+			IdentityDeepCopyJSON,
+			IdentityAssignTypeName,
+		},
+	}
+}
+
+// AddConversionHandlers adds the specified conversion handlers to the end of the conversion list.
+func (builder *ConversionFunctionBuilder) AddConversionHandlers(conversionHandlers ...ConversionHandler) {
+	builder.conversions = append(builder.conversions, conversionHandlers...)
+}
+
+// PrependConversionHandlers adds the specified conversion handlers to the beginning of the conversion list.
+func (builder *ConversionFunctionBuilder) PrependConversionHandlers(conversionHandlers ...ConversionHandler) {
+	builder.conversions = append(conversionHandlers, builder.conversions...)
+}
+
+func (builder *ConversionFunctionBuilder) BuildConversion(params ConversionParameters) []dst.Stmt {
+	for _, conversion := range builder.conversions {
+		result := conversion(builder, params)
+		if len(result) > 0 {
+			return result
+		}
+	}
+
+	panic(fmt.Sprintf("don't know how to perform conversion for %T -> %T", params.SourceType, params.DestinationType))
+}
+
+// IdentityConvertComplexOptionalProperty handles conversion for optional properties with complex elements
+// This function generates code that looks like this:
+// 	if <source> != nil {
+//		<code for producing result from destinationType.Element()>
+//		<destination> = &<result>
+//	}
+func IdentityConvertComplexOptionalProperty(builder *ConversionFunctionBuilder, params ConversionParameters) []dst.Stmt {
+	destinationType, ok := params.DestinationType.(*OptionalType)
+	if !ok {
+		return nil
+	}
+
+	sourceType, ok := params.SourceType.(*OptionalType)
+	if !ok {
+		return nil
+	}
+
+	// TODO: Dislike Typed suffix here -- change
+	tempVarIdent := builder.IdFactory.CreateIdentifier(params.NameHint+"Typed", NotExported)
+
+	newSource := &dst.UnaryExpr{
+		X:  params.ClonedSource(),
+		Op: token.MUL,
+	}
+
+	innerStatements := builder.BuildConversion(
+		params.WithDestination(dst.NewIdent(tempVarIdent)).
+			WithDestinationType(destinationType.Element()).
+			withAdditionalConversionContext(destinationType).
+			WithAssignmentHandler(AssignmentHandlerDefine).
+			WithSource(newSource).
+			WithSourceType(sourceType.Element()))
+
+	// Tack on the final assignment
+	innerStatements = append(
+		innerStatements,
+		astbuilder.SimpleAssignment(
+			params.ClonedDestination(),
+			token.ASSIGN,
+			&dst.UnaryExpr{
```

Comment:
Use `astbuilder.AddrOf()`

Reply:
Done - fixed other usages of UnaryExpr too